### PR TITLE
Use TAP with alma - initial version

### PR DESCRIPTION
**Pull request by andamian**

### CODE REVIEW

File: astroquery/alma/core.py [hunk 1]
```diff
@@ -26,9 +26,69 @@
...
+# Map from ALMA ObsCore result to ALMA original query result
+# The map is provided in order to preserve the API of the alma module
+_OBSCORE_TO_ALMARESULT = {
+    'proposal_id': 'Project code',
+    's_ra':	'RA',
+    's_dec': 'Dec',
+    'gal_longitude': 'Galactic longitude',
+    'gal_latitude':	'Galactic latitude',
+    'band_list': 'Band',
+    's_region':	'Spatial resolution',
```
Comment (keflavich) [inline astroquery/alma/core.py]: could use some whitespace cleanup

File: astroquery/alma/core.py [hunk 2]
```diff
@@ -26,9 +26,69 @@
 from ..utils.process_asyncs import async_to_sync
 from ..query import QueryWithLogin
 from . import conf, auth_urls
+try:
+    import pyvo
+except ImportError:
+    print('Please install pyvo. astropy.vo does not work without it.')
```
Comment (keflavich) [inline astroquery/alma/core.py]: as w/previous: shouldn't this just raise?

File: astroquery/alma/core.py [hunk 3]
```diff
@@ -26,9 +26,69 @@
...
+    's_fov': 'Field of view',
+    #  TODO 'spatial_scale_max': 'Largest angular scale',
+    'qa2_passed': 'QA2 Status',
+    #  TODO	COUNT
+    'science_keyword': 'Science keyword',
+    'scientific_category': 'Scientific category'
+}
+
+ALMA_BANDS = {
+    '3': (84*10**9*u.Hz, 116*10**9*u.Hz),
```
Comment (keflavich) [inline astroquery/alma/core.py]: you don't like GHz?

File: astroquery/alma/core.py [hunk 4]
```diff
@@ -124,95 +187,88 @@ def query_async(self, payload, cache=True, public=True, science=True,
...
+        # ALMA query parameters. Transform them all in SIAv2 paramters
+
+        cal_level = [2, 3] if science else None
+        #  TODO public data only
+
+        # coordinates can be expressed as pos (SIAv2), rad_dec (ICRS coord),
+        # galactic (Galactic longitude, Galactic latitude str) or
+        # source_name_resolver (object name) or combination.
+        # Add them all to the pos list
+        radius = payload.get('radius', 0.016666666666667*u.deg)
```
Comment (keflavich) [inline astroquery/alma/core.py]: maybe `1*u.arcmin`?

**Comment by keflavich** [comment]
To give you a sense of the API as I've used it, here's a list of a bunch of different queries I have run:
```
Alma.query_region(coordinate=orionkl_coords, radius=4*u.arcmin, public=False, science=False)
Alma.query_object('M83', public=False, science=False)
Alma.query(payload={'pi_name':'Bally'}, public=False)
alma.query(payload=dict(project_code='2016.1.00165.S'), public=False, cache=False)
alma.query(payload=dict(project_code='2017.1.01355.L', source_name_alma='G008.67'),)
Alma.query_region(coordinates.SkyCoord('5:35:14.461 -5:21:54.41', frame='fk5', unit=(u.hour, u.deg)), radius=0.034*u.deg)
Alma.query_region(coordinates.SkyCoord('5:35:14.461 -5:21:54.41', frame='fk5', unit=(u.hour, u.deg)), radius=0.034*u.deg, payload={'energy.frequency-asu':'215 .. 220'})
Alma.query(payload=dict(project_code='2012.*', public_data=True))
Alma.query(payload={'frequency':'96 .. 96.5'}, cache=False)
rslt = Alma.query_object('M83', band_list=[4,5,8])
Alma.query(payload={'pi_name':'ginsburg', 'band_list':'6'})
```

but I'll note that my history of using the band & frequency queries is littered with dozens of failed attempts to get the syntax right.  The API we'd like to support does not have to take the same syntax, and in fact it would be helpful if we could check the syntax (and specify it) user-side.  

So for the basic query API, as long as we continue to support the basic region queries and some variants of the criteria query (on PI, project ID, frequency, band, polarization [tricky, I know], etc), it's OK.  I think we'll have to live with syntax-breaking in the uploaded values (e.g., 96 .. 96.5 above) because we don't really know what the old syntax was anyway.

**Comment by andamian** [comment]
The majority of those examples should work with the changes in this PR (and probably a few others). Wild cards (project_code='2012.*') however do not work.
That being said I suggest the following:
Have query interface for general users and one for advanced ALMA users. Thi,. in fact, is Simple Image Access and TAP Obscore protocols' raison d'être.
More concrete, I suggest exposing the SIA attributes in the `query*` methods. Issue deprecate warnings for `payload` arguments but for now try to have them still work by "translating" them into SIA constraints whenever possible. Same deal for the results. This will constitute the general user ALMA API.

For users with more intimate knowledge about ALMA, we can create a method to query the TAP service. Users will be given tools to explore the schema of the TAP service (`help_tap`) and write their own queries. That's where they can use wildcards, sorting, grouping and all the other goodies.

This approach is not too complicated, keeps the library simple, tries to make the changes less disruptive and appeals to any type of audience, including power users.

In summary, I'm proposing:
1. add `sia` argument to `query*` as a dictionary with the following expected keys: `pos, band, time, pol, field_of_view, spatial_resolution, spectral_resolving_power, exptime, timeres, publisher_did, facility, collection, instrument, data_type, calib_level, target_name, res_format, maxrec`. Or should they be exposed directly as method arguments the way I've proposed it in the `vo` module?
2. Add deprecate warnings when `payload` argument is detected but attempt to perform operations when possible.
3. Add `help_sia` method. Also update `help` method to explain that the interface is deprecated and which arguments are still supported in the current version.
- Add `query_tap` method to query `ObsCore` via `TAP`.
- Add `help_tap` to query ALMA TAP schema and return information regarding the available `ObsCore` columns and their types.

Sounds acceptable?

**Comment by keflavich** [comment]
I like this plan.

Re: (1) - let's have these as options to `query_*`.  I don't immediately understand how they would work as methods, but we want to continue supporting `query*` interfaces since the rest of astroquery has them.

Everything else is fine.

**Comment by andamian** [comment]
> I like this plan.
> 
> Re: (1) - let's have these as options to `query_*`. I don't immediately understand how they would work as methods, but we want to continue supporting `query*` interfaces since the rest of astroquery has them.
> 
> Everything else is fine.

What I meant is that `query*` can either have all those arguments add with defaults or use a single `sia` generic dictionary argument (similar to `payload`). The former is a bit better because works with autocompletion and IDEs, while the latter is consistent with the original `payload` (generic dictionary argument + corresponding `help` method).

**Comment by keflavich** [comment]
yeah, I like the IDE-friendlier specific arguments approach.

**Comment by andamian** [comment]
I'm adding to the thread the link to the google docs used for mapping between old parameters and the new ones. Needs to be reviewed and commented on.
https://docs.google.com/spreadsheets/d/1X0rrHuTMtU4Tp_CnBR1CdylffvNUIliTgdNMhnLCCws/edit#gid=0

**Comment by andamian** [comment]
The query samples you've posted with the new API:
```
        alma = Alma()
        #alma.query_region(coordinate=orionkl_coords, radius=4 * u.arcmin,
        #                  public=False, science=False)
        result = alma.query_object('M83', public=False, science=False)
        assert len(result) > 0
        #TODO test public
        result = alma.query_tap(
            "select * from ivoa.ObsCore where obs_creator_name like '%Bally%'")
        assert result
        for row in result:
            assert 'Bally' in row['obs_creator_name']
        result = alma.query_tap\
            ("select * from ivoa.Obscore where proposal_id='2016.1.00165.S'")
        assert result
        for row in result:
            assert '2016.1.00165.S' == row['proposal_id']
        result = alma.query_tap(
            "select * from ivoa.Obscore where proposal_id='2017.1.01355.L' "
            "and target_name like '%G008.67%' and data_rights='Public'")
        assert result
        for row in result:
            assert '2017.1.01355.L' == row['proposal_id']
            assert 'Public' == row['data_rights']
            assert 'G008.67'in row['target_name']

        result = alma.query_region(
            coordinates.SkyCoord('5:35:14.461 -5:21:54.41', frame='fk5',
                                 unit=(u.hour, u.deg)), radius=0.034 * u.deg)
        assert result

        # result = alma.query_region(
        #     coordinates.SkyCoord('5:35:14.461 -5:21:54.41', frame='fk5',
        #                          unit=(u.hour, u.deg)), radius=0.034 * u.deg,
        #     payload={'energy.frequency-asu': '215 .. 220'})
        #
        result = alma.query_tap("select * from ivoa.ObsCore where proposal_id "
                                "like '2012.%' and data_rights='Public'")
        assert result
        for row in result:
            assert '2012.' in row['proposal_id']
            assert 'Public' == row['data_rights']

        result = alma.query(payload={'frequency': '96 .. 96.5'})
        assert result
        for row in result:
            # TODO not sure how to test this
            pass

        result = alma.query_object('M83', band_list=[4, 5, 8])
        assert result
        for row in result:
            assert row['band_list'] in ['4', '5', '8']

        result = alma.query_tap(
            "select * from ivoa.ObsCore where lower(obs_creator_name) like "
            "'%ginsburg%' and band_list = '6'")
        assert result
        for row in result:
            assert '6' == row['band_list']
            assert 'ginsburg' in row['obs_creator_name'].lower()
```

Looks OK? I didn't know what `energy.frequency-asu` is or how to find it in ALMA ObsCore.

**Comment by keflavich** [comment]
Yes, looks good, though I definitely want a shortcut approach to those SQL queries!

I don't know what `energy.frequency-asu` is either.  It looks like I was just trying to find a way to search within a frequency range.

You can put some of this stuff in as actual tests and I can give you some inline comments

**Comment by keflavich** [comment]
We've been given a deadline of "this Wednesday", after which the present ALMA archive will be disabled.

There are problems.

All queries were failing with:
```
rslt = Alma.query_object('W51', pi_name='*Ginsburg*', public=False)
Traceback (most recent call last):
  File "<ipython-input-12-dbcc322179c1>", line 1, in <module>
    rslt = Alma.query_object('W51', pi_name='*Ginsburg*', public=False)
  File "/home/adam/repos/astroquery/astroquery/utils/class_or_instance.py", line 25, in f
    return self.fn(obj, *args, **kwds)
  File "/home/adam/repos/astroquery/astroquery/utils/process_asyncs.py", line 26, in newmethod
    response = getattr(self, async_method_name)(*args, **kwargs)
  File "/home/adam/repos/astroquery/astroquery/alma/core.py", line 245, in query_object_async
    payload=payload, **kwargs)
  File "/home/adam/repos/astroquery/astroquery/alma/core.py", line 343, in query_async
    result[_OBSCORE_TO_ALMARESULT[ii]] = result[ii]
  File "/home/adam/repos/astropy/astropy/table/table.py", line 1648, in __getitem__
    return self.columns[item]
  File "/home/adam/repos/astropy/astropy/table/table.py", line 239, in __getitem__
    return OrderedDict.__getitem__(self, item)
KeyError: 'frequency'
```
which resulted from a bad mapping..

File: astroquery/alma/core.py [hunk 5]
```diff
@@ -79,140 +255,167 @@ def query_region_async(self, coordinate, radius, cache=True, public=True,
...
-        Use the ALMA query validator service to check whether the keywords are
-        valid
+            return result
+        # add 'Observation date' column
+        result['Observation date'] = \
+            [Time(_['t_min'], format='mjd').strftime(ALMA_DATE_FORMAT)
+                for _ in result]
+        for ii in _OBSCORE_TO_ALMARESULT:
+            if _OBSCORE_TO_ALMARESULT[ii] not in result.columns:
+                # duplicate equivalent columns
```
Comment (keflavich) [inline astroquery/alma/core.py]: why are we doing this?   For backward compatibility?  I don't like it, I'd rather toss one or the other.  I guess we can toss the old, stick with the new.  We should add, though, a renaming utility using the `_OBSCORE_TO_ALMARESULT` mapping to allow users to easily recover the old naming scheme.

File: astroquery/alma/tests/test_alma.py [hunk 6]
```diff
@@ -14,100 +22,333 @@ def data_path(filename):
...
+    tap_mock.search.assert_called_once_with(
+        "select * from ivoa.obscore WHERE "
+        "(INTERSECTS(CIRCLE('ICRS',1.0,2.0,0.16666666666666666), s_region) = 1) "
+        "AND band_list LIKE '%3%' AND calib_level>1 AND data_rights='Proprietary'",
+        band_list=[3], language='ADQL')
+    # check ObsCore columns in result are the same with old ones
+    row = result[0]
+    for item in _OBSCORE_TO_ALMARESULT.items():
+        if item[0] == 't_min':
+            assert Time(row[item[0]], format='mjd').strftime('%d-%m-%Y') ==\
```
Comment (keflavich) [inline astroquery/alma/tests/test_alma.py:278]: my changes cause a failure here, but I suspect that's because the data are out of date - alma_onerow contains data that aren't returned from alma at present.

File: astroquery/alma/core.py [hunk 7]
```diff
@@ -38,37 +191,60 @@ class AlmaClass(QueryWithLogin):
...
+        Query the archive for a source name.
 
         Parameters
         ----------
         object_name : str
-            The object name.  Will be parsed by SESAME on the ALMA servers.
-        cache : bool
-            Cache the query?
+            The object name.  Will be resolved by astropy.coord.SkyCoord
+        cache : deprecated
```
Comment (keflavich) [inline astroquery/alma/core.py:232]: is it necessary to deprecate caching?  Does pyvo not support it?  The caching functionality is extremely useful, even though it is problematic in several ways (we have no way to test if the cache is out of date, for instance)

File: astroquery/alma/core.py [hunk 8]
```diff
@@ -8,28 +8,181 @@
...
+    'gal_longitude': 'Galactic longitude',
+    'gal_latitude': 'Galactic latitude',
+    'band_list': 'Band',
+    's_region': 'Spatial resolution',
+    'em_resolution': 'Frequency resolution',
+    'antenna_arrays': 'Array',
+    'is_mosaic': 'Mosaic',
+    't_exptime': 'Integration',
+    'obs_release_date': 'Release date',
+    'frequency': 'Frequency support',
```
Comment (keflavich) [inline astroquery/alma/core.py]: I changed this & coi_name, but that triggered some remote test failures.  My changes may therefore be incorrect and need to be reverted - feel free to do so, but let's make sure the docs examples work.

File: CHANGES.rst [hunk 9]
```diff
@@ -10,6 +10,24 @@ esa/hubble
...
+- The Splatalogue ID querying is now properly cached in the `astropy` cache
+  directory (Issue [#423]) The scraping function has also been updated to reflect
+  the Splatalogue webpage. [#1772]
+
+UKIDSS
+^^^^^^
+
+- Updated to ``UKIDSSDR11PLUS`` as the default version [#1767]
+
+
```
Comment (keflavich) [inline CHANGES.rst:30]: somehow these changelog entries have been added - they're not part of this PR.  Maybe they got unintentionally moved here during the squash?

Comment (ceb8) [reply CHANGES.rst:30]: It seems like at least some if not all of the actual changes associated with these changelog entries have also been added.

Comment (andamian) [reply CHANGES.rst]: I have no idea how they got in and where they are coming from. For me, re-base is always a nightmare. And I'm not sure about the benefits.

Comment (andamian) [reply CHANGES.rst]: Can someone tell me why it thinks this PR added those changes when they are already in master?

Comment (keflavich) [reply CHANGES.rst]: are they maybe duplicated in different sections here?

Comment (andamian) [reply CHANGES.rst]: `github` flagged a conflict with `master`. Fixed it and the extra files are now gone. Please double-check.

Comment (andamian) [reply CHANGES.rst]: Can someone assign it to the next milestone to make sure I'm not missing anything? Thanks

Comment (bsipocz) [reply CHANGES.rst]: At the time of tagged release the changelog gets a restructure, I think that has tricked your git to think you have all these as new additions. Anyway, it's all been cleared out now.

**Comment by bsipocz** [comment]
Thanks @andamian!
